src/buildblock/ProjData.cxx

```diff
@@ -225,6 +229,27 @@ read_from_file(const string& filename,
 //  this->exam_info_sptr.reset(new ExamInfo(new_exam_info));
 //}
 
+
+ProjDataInMemory
```

Comment:
Frankly, I do not like ABC returning an instance of derived class - I firmly believe that ABC should not be aware about classes derived from it.
@ashgillman How about this trick that we use in SIRF: add this method to `ProjDataInMemory`:
```
virtual ProjData* same_proj_data
	(stir::shared_ptr<const stir::ExamInfo> sptr_exam_info,
	stir::shared_ptr<stir::ProjDataInfo> sptr_proj_data_info) const
{
	ProjData* ptr_pd =
		new ProjDataInMemory(sptr_exam_info, sptr_proj_data_info);
	return ptr_pd;
}
```
and same-prototype one to `ProjData`, returning zero pointer for now (until all other derived classes get proper ones eventually). It can then be used instead of what you have on line 238 to create `unique_ptr<ProjData>` to be returned by your `get_subset`.

Reply:
Seems that STIR indeed misses functionality like `same_proj_data`. We have `get_empty_copy` in the image front, but not for `ProjData`. However, this would be a considerable amount of work. In fact, it is virtually impossibly to create a GE/Siemens/whatever proj-data from scratch as we don't have all of the info on the `ExamInfo` and `ProjDataInfo` that would be needed for that. The usual strategy for that is to copy existing files, and then modify the things we need. Implementing even that is a large amount of work. We can currently only create new `ProjDataInterfile` and `ProjDataInMemory` objects.

It might even be entirely undesirable to have it return a `ProjData` of the same type.

Because of all this we took the past of least resistance at the moment. we could easily decide to return a `unique_ptr<ProjData>` and optional arguments such as a a filename, and `ios:openmode` flags in which case it will create a `ProjDataInterfile` accordingly. More than that I think is very much out-of-scope

Reply:
@KrisThielemans the most important thing here is to return `unique_ptr<ProjData>` whichever way the subset object is created, otherwise it will not be possible to wrap it into `PETAcquisitionData` without cloning (one copy already being created at present)

Reply:
@KrisThielemans sorry to be such a bore but I tried this in VS 2019:
```
#include <iostream>
#include <memory>

class A {
public:
	int i;
	A(int v) : i(v) {}
	A(const A& a)
	{
		std::cout << "copy constructor called\n";
		i = a.i;
	}
};

A f()
{
	A a(8);
	return a;
}

int main()
{
	A a(42);
	std::cout << "calling f()...\n";
	a = f();
	std::cout << "calling f() with make_shared...\n";
	auto b = std::make_shared<A>(f());
	return 0;
}
```
and I got
```
calling f()...
copy constructor called
calling f() with make_shared...
copy constructor called
copy constructor called
```
On VM with Ubuntu20.04 and g++ version 9.3.0, I get
```
calling f()...
calling f() with make_shared...
copy constructor called
```
So 
* apparently VS 2019 is not clever enough to avoid copying return value, and 
* it looks like
```
make_shared<A>(a)
```
actually simply does this (as I always thought):
```
shared_ptr<A>(new A(a))
```
So, with `ProjData::get_subset`, why not take this IMHO simplest possible way to avoid unnecessary copies:
```
unique_ptr<ProjData>
ProjData::get_subset(const std::vector<int>& views) const {
...
auto uptr_subset_proj_data = unique_ptr<ProjData>(new ProjDataInMemory(exam_info_sptr, subset_proj_data_info_sptr));
subset_proj_data *uptr_subset_proj_data;
...
return uptr_subset_proj_data;
}
```

Reply:
interesting. makes me wonder what Microsoft plans are for their C++ compiler if they even didn't do that. thanks for trying!  please do change the code then...

Reply:
@KrisThielemans I have just forked Ashley's STIR branch (could not push to it) but it does not have your latest edits - where can I get them from? UCL/STIR does not seem to have `subset_object` branch.

Reply:
@KrisThielemans forget about my previous comment - I opened a file before forking and forgot that leafpad does not automatically update edited file.

Reply:
I get this build error:
```
/home/sirfuser/devel/build/sources/STIR/src/recon_buildblock/DataSymmetriesForBins_PET_CartesianGrid.cxx:141:88: error: ‘using element_type = const class stir::ProjDataInfo’ {aka ‘const class stir::ProjDataInfo’} has no member named ‘get_org_proj_data_info_sptr’; did you mean ‘construct_proj_data_info’?

if(!is_null_ptr(dynamic_cast<const ProjDataInfoCylindrical*>(proj_data_info_ptr->get_org_proj_data_info_sptr().get())))
```
Indeed, `proj_data_info_ptr` is declared as `ProjDataInfo`, which does not have that new method. It is cast to `ProjDataInfoSubsetByView` in the quoted line
```
  if(!is_null_ptr(dynamic_cast<const ProjDataInfoSubsetByView *>(proj_data_info_ptr.get())))
```
but the result of the cast is not assigned to any variable.
I edited as follows
```
  const ProjDataInfoSubsetByView *ptr=dynamic_cast<const ProjDataInfoSubsetByView *>(proj_data_info_ptr.get());
  if(!is_null_ptr(ptr))
```
and the error message disappeared.
@KrisThielemans But I thought you already managed to build ok, and `git fetch` does not fetch - were you working on some other fork/branch?

Reply:
this was all done.